### PR TITLE
rl-2311: Add release notes on lib

### PR DESCRIPTION
**Pull request by infinisil**
## Description of changes

NixOS releases are also `lib` releases :).
As the main `lib` maintainer, here are the release notes for it.
These were collected from looking at the `git diff` since 22.11.

I boldly inserted them at the very front, because compared to all the package updates, `lib` may contain functions that everybody uses.

This work is sponsored by [Antithesis](https://antithesis.com/) :sparkles:

## Things done

- [x] Built the manual

### CODE REVIEW

**Comment by infinisil** [comment]
I notably didn't include these notes on `lib.systems`, because I'm not maintaining that and barely understand it:

### Platforms
- Added `lib.systems.equals`
- GCC feature support for architectures `x86-64`, `x86-64-v2`, `x86-64-v3`, `x86-64-v4`, `nehalem`, `alderlake` and `znver4`
- Better support for Rust cross compilation (https://github.com/NixOS/nixpkgs/pull/233628)
- New elaborated platform properties:
  - `hasSharedLibraries`
  - `ubootArch`
- New doubles:
  - "mips-linux" "mips64-linux" "mips-none" "mips64-none"
- New platforms:
  - bluefield2
- New examples:
  - bluefield2
  - gnu64_simplekernel
  - mips64-embedded
  - mips-embedded
- Removed examples:
  - mipsisa32r6-linux-gnu
  - mipsisa32r6el-linux-gnu
  - mipsisa64r6-linux-gnuabin32
  - mipsisa64r6el-linux-gnuabin32
  - mipsisa64r6-linux-gnuabi64
  - mipsisa64r6el-linux-gnuabi64

Ping @alyssais @Ericson2314 @amjoseph-nixpkgs

File: nixos/doc/manual/release-notes/rl-2311.section.md
Comment (infinisil) [inline nixos/doc/manual/release-notes/rl-2311.section.md:91]: Ping @pennae

**Review by alejandrosame — DISMISSED** [review]
LGTM, just please move the whole block to the end of the file and we'll edit it appropriately with nixpkgs internal section afterwards.

**Comment by infinisil** [comment]
Moved it to the bottom for now, but these `lib` changes should _not_ go under "internal" changes. They are a stable and exposed API. I also think that generally curated release notes should go over non-curated ones.

File: nixos/doc/manual/release-notes/rl-2311.section.md
Comment (roberth) [inline nixos/doc/manual/release-notes/rl-2311.section.md:661]:
```suggestion
  Nixpkgs accepts PRs that set `meta.mainProgram` on packages where it makes sense.
  Use `lib.meta.getExe' pkg "some-command"` to avoid the warning and/or select a different executable.
```

Ideally we have better data: more mainPrograms.
Nits:
- commands tend to be lower snake case
- avoid "binary" because scripts aren't binaries (`/bin` is very very unfortunate)

Comment (roberth) [reply nixos/doc/manual/release-notes/rl-2311.section.md:661]: `lib.meta.getExe` is annoying. Do we really want people to type that?
Sublibraries are cool, but everyone writes `lib.getExe`.

File: nixos/doc/manual/release-notes/rl-2311.section.md
Comment (roberth) [inline nixos/doc/manual/release-notes/rl-2311.section.md:663]: This should be in a section of its own, because the functions did not change and mentioning it serves a different purpose.

**Review by roberth — DISMISSED** [review]

File: nixos/doc/manual/release-notes/rl-2311.section.md
Comment (infinisil) [inline nixos/doc/manual/release-notes/rl-2311.section.md:607]: @roberth I restructured it into just two sections instead now, one of them is here, and includes both new stuff but also improvements to stuff (intentionally not being specific what "stuff" is, so it can also include docs)

**Comment by infinisil** [comment]
Ping @roberth, does it look good now?

Also @alejandrosame, do you want to merge this yourself or can I do that myself? (and then backport)

**Comment by alejandrosame** [comment]
> Ping @roberth, does it look good now?
> 
> Also @alejandrosame, do you want to merge this yourself or can I do that myself? (and then backport)

Please, merge it. I'm away for quite some time today (plus, not sure I really have merge rights for this)

File: nixos/doc/manual/release-notes/rl-2311.section.md
Comment (roberth) [inline nixos/doc/manual/release-notes/rl-2311.section.md:698]:
```suggestion
  `lib.options.literalDocBook` has been removed after deprecation in 22.11.
```

File: nixos/doc/manual/release-notes/rl-2311.section.md
Comment (roberth) [inline nixos/doc/manual/release-notes/rl-2311.section.md:684]: Breaking changes should go first, not last.

File: nixos/doc/manual/release-notes/rl-2311.section.md
Comment (roberth) [inline nixos/doc/manual/release-notes/rl-2311.section.md:688]: Changes with a proper migration path (deprecations in logs, as well as good error messages) should be demoted to a section that almost all readers can skip, or just omitted.

**Review by roberth — COMMENTED** [review]
Prioritization and a typo. LGTM after that.

**Comment by infinisil** [comment]
@roberth Done! [(diff)](https://github.com/NixOS/nixpkgs/compare/69549dfb1eb818d3f62a66b56ce63e3e6265c0c6..38664f70d14626d7bc44a159a075ce5202824665)

**Review by roberth — APPROVED** [review]
Awesome.

**Pull request merged**
